### PR TITLE
chore: implement deleting custom roles

### PR DESCRIPTION
**Pull request by Emyrk**
When deleting a custom role, the role can still be assigned to a user.

### CODE REVIEW

**Comment by Emyrk** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/coder/coder/14101?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14101** <a href="https://app.graphite.dev/github/pr/coder/coder/14101?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#14099** <a href="https://app.graphite.dev/github/pr/coder/coder/14099?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Emyrk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: coderd/apidoc/docs.go [hunk 1]
```diff
@@ -2549,6 +2549,51 @@ const docTemplate = `{
                 }
             }
         },
+        "/organizations/{organization}/members/roles{roleName}": {
```
Comment (jaaydenh) [inline coderd/apidoc/docs.go:2552]: missing a `/` after roles?

```suggestion
        "/organizations/{organization}/members/roles/{roleName}": {
```

Comment (Emyrk) [reply coderd/apidoc/docs.go:2552]: Yes! :+1: 

Typo in my swagger comment

File: enterprise/coderd/roles.go [hunk 2]
```diff
@@ -131,6 +132,85 @@ func (api *API) patchOrgRoles(rw http.ResponseWriter, r *http.Request) {
...
+	rolename := chi.URLParam(r, "roleName")
+	roles, err := api.Database.CustomRoles(ctx, database.CustomRolesParams{
+		LookupRoles: []database.NameOrganizationPair{
+			{
+				Name:           rolename,
+				OrganizationID: organization.ID,
+			},
+		},
+		ExcludeOrgRoles: false,
+		OrganizationID:  organization.ID,
```
Comment (aslilac) [inline enterprise/coderd/roles.go:170]: weird duplication of `OrganizationID` 🤔

File: enterprise/coderd/roles.go [hunk 3]
```diff
@@ -131,6 +132,85 @@ func (api *API) patchOrgRoles(rw http.ResponseWriter, r *http.Request) {
...
+	})
+	if httpapi.IsUnauthorizedError(err) {
+		httpapi.Forbidden(rw)
+		return
+	}
+	if err != nil {
+		httpapi.InternalServerError(rw, err)
+		return
+	}
+	aReq.New = database.CustomRole{}
```
Comment (aslilac) [inline enterprise/coderd/roles.go:209]: clever 😄

**Review by aslilac — APPROVED** [review]
lgtm!

File: enterprise/coderd/roles.go (hunk 2 above)
Comment (Emyrk) [reply enterprise/coderd/roles.go:170]: Yea, it actually is not necessary, but we have a linter that makes you fill out all fields in a struct. This felt more intuitive that doing `uuid.Nil`. But you have to do something, or the linter yells at you :cry:

File: coderd/database/migrations/000240_delete_user_roles.up.sql [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+-- Attach the function to deleting the custom role
+CREATE TRIGGER remove_organization_member_custom_role
+	BEFORE DELETE ON custom_roles FOR EACH ROW
+	EXECUTE PROCEDURE remove_organization_member_role();
+
+
+COMMENT ON TRIGGER
+	remove_organization_member_custom_role
+	ON custom_roles IS
+		'When a custom_role is deleted, this trigger removes the role from all organization members.';
```
Comment (f0ssel) [inline coderd/database/migrations/000240_delete_user_roles.up.sql:35]: I generally prefer this logic to live in app code in some sort of helper function, probably in a tx. I don't think this has to be changed here since I'm not sure how the team at large feels about this kind of design, but wanted to point out that I think it's easier to follow execution flow and also easily update the behavior as well since changing this behavior now requires knowledge of sql triggers. 

I see the benefits from a data integrity standpoint. Does this serve as some sort of performance advantage as well?

File: enterprise/coderd/roles.go (hunk 2 above)
Comment (f0ssel) [reply enterprise/coderd/roles.go:170]: tldr on why this field can't be removed?

File: enterprise/coderd/roles_test.go [hunk 5]
```diff
@@ -310,6 +310,134 @@ func TestCustomOrganizationRole(t *testing.T) {
...
+		// could not be assigned roles anymore.
+		//nolint:gocritic // org_admin cannot assign to themselves
+		_, err = owner.UpdateOrganizationMemberRoles(ctx, first.OrganizationID, orgAdminUser.ID.String(), codersdk.UpdateRoles{
+			Roles: []string{rbac.ScopedRoleOrgAdmin(first.OrganizationID).Name},
+		})
+		require.NoError(t, err)
+	})
+
+	// Verify deleting a custom role cascades to all members
+	t.Run("DeleteRoleCascadeMembers", func(t *testing.T) {
```
Comment (f0ssel) [inline enterprise/coderd/roles_test.go:374]: 👍

**Review by f0ssel — APPROVED** [review]

File: coderd/database/dbauthz/dbauthz.go [hunk 6]
```diff
@@ -958,6 +958,20 @@ func (q *querier) DeleteCoordinator(ctx context.Context, id uuid.UUID) error {
 	return q.db.DeleteCoordinator(ctx, id)
 }
 
+func (q *querier) DeleteCustomRole(ctx context.Context, arg database.DeleteCustomRoleParams) error {
+	if arg.OrganizationID.UUID != uuid.Nil {
+		if err := q.authorizeContext(ctx, policy.ActionDelete, rbac.ResourceAssignOrgRole.InOrg(arg.OrganizationID.UUID)); err != nil {
+			return err
+		}
+	} else {
+		if err := q.authorizeContext(ctx, policy.ActionCreate, rbac.ResourceAssignRole); err != nil {
```
[the remark this answers is not in the document]
Comment (Emyrk) [reply coderd/database/dbauthz/dbauthz.go:967]: Great catch, I honestly just missed it. I'll update the tests to enforce this.

File: coderd/database/dbauthz/dbauthz_test.go [hunk 7]
```diff
@@ -1247,6 +1247,19 @@ func (s *MethodTestSuite) TestUser() {
...
+				UUID:  uuid.New(),
+				Valid: true,
+			},
+		})
+		check.Args(database.DeleteCustomRoleParams{
+			Name:           customRole.Name,
+			OrganizationID: customRole.OrganizationID,
+		}).Asserts(
+			rbac.ResourceAssignOrgRole.InOrg(customRole.OrganizationID.UUID), policy.ActionDelete)
+	}))
```
[the remark this answers is not in the document]
Comment (Emyrk) [reply coderd/database/dbauthz/dbauthz_test.go:1262]: Yup, just added :+1:. Saw your catch above

File: coderd/database/migrations/000240_delete_user_roles.up.sql (hunk 4 above)
Comment (Emyrk) [reply coderd/database/migrations/000240_delete_user_roles.up.sql:35]: Yea, I generally agree with you. In this case the data integrity is a security critical thing though.

Currently when making a new custom_role, you have to have a superset of the permissions to make it. So you cannot make a role that has greater privilege than yourself. 

You could imagine a scenario where a role is created with a small set of perms, then deleted. Then a new role is created with the same name that has more permissions, and because a user had the old role, they are granted this new one.

At present, this would be less malicious, and more likely a mistake based on our roles, but the problem still remains.

----

So essentially, I choose the trigger as a security guarantee. We have some prior art for this for deleting [user api keys](https://github.com/coder/coder/blob/ab9e708bec9a8a582f6912e2563f9dfae3c6f015/coderd/database/migrations/000120_trigger_delete_user_apikey.up.sql#L26-L30).

A golang helper function just feels less safe imo.

**Comment by Emyrk** [comment]
> tldr on why this field can't be removed?

We have a linter that requires all struct fields be explicitly set. So you could put `uuid.Nil` and it would still work, just felt like it might confuse someone else to see a `uuid.Nil` :man_shrugging:
